packages/gateway/src/gateway.js

```diff
@@ -0,0 +1,133 @@
+/* eslint-env serviceworker, browser */
+
+import pAny from 'p-any'
+import pSettle from 'p-settle'
+
+import { getCidFromSubdomainUrl } from './utils/cid.js'
+
+/**
+ * @typedef {Object} GatewayResponse
+ * @property {Response} response
+ * @property {string} url
+ * @property {number} responseTime
+ */
+
+/**
+ * Handle gateway request
+ *
+ * @param {Request} request
+ * @param {import('./env').Env} env
+ * @param {import('./index').Ctx} ctx
+ */
+export async function gatewayGet(request, env, ctx) {
+  const reqUrl = new URL(request.url)
+  const cid = getCidFromSubdomainUrl(reqUrl)
+
+  const gatewayReqs = env.ipfsGateways.map(async (url, index) => {
+    const ipfsUrl = new URL('ipfs', url)
+    const controller = new AbortController()
+    const startTs = Date.now()
+    const timer = setTimeout(() => controller.abort(), env.REQUEST_TIMEOUT)
+
+    const response = await fetch(
+      `${ipfsUrl.toString()}/${cid}${reqUrl.pathname || ''}`,
+      { signal: controller.signal }
+    )
+
+    clearTimeout(timer)
+    /** @type {GatewayResponse} */
+    const gwResponse = {
+      response,
+      url,
+      responseTime: Date.now() - startTs,
+    }
+    return gwResponse
+  })
+
+  try {
+    const { response, url } = await pAny(gatewayReqs)
+
+    ctx.waitUntil(
+      (async () => {
+        const responses = await pSettle(gatewayReqs)
+        const successFullResponses = responses.filter(
+          (r) => r.value?.response?.ok
+        )
+
+        await Promise.all([
+          updateMetrics(request, env, responses, url),
```

Comment:
We discussed having a separate metrics instance per gateway so that we can update each one as soon as it finishes rather than wait for all of them to finish. It also simplifies the logic in that class significantly.

@vasco-santos to decide if done here or in a followup PR.